pkg/operator/k8sutil/pod.go

```diff
@@ -120,46 +120,59 @@ func SetPodVersion(pod *v1.Pod, key, version string) {
 	pod.Annotations[key] = version
 }
 
-// DeleteDaemonDeployment makes a best effort at deleting the daemon pods, including the deployment and daemonset
-func DeleteDaemonDeployment(context *clusterd.Context, namespace, name string) {
-	logger.Infof("removing %s pods if they exist", name)
+// DeleteDeployment makes a best effort at deleting a deployment and its pods, then waits for them to be deleted
+func DeleteDeployment(context *clusterd.Context, namespace, name string) {
+	logger.Infof("removing %s deployment if it exists", name)
+	deleteAction := func(name string, options *metav1.DeleteOptions) error {
+		return context.Clientset.ExtensionsV1beta1().Deployments(namespace).Delete(name, options)
+	}
+	getAction := func(name string) error {
+		_, err := context.Clientset.ExtensionsV1beta1().Deployments(namespace).Get(name, metav1.GetOptions{})
+		return err
+	}
+	deletePodsAndWait(context, namespace, name, deleteAction, getAction)
+}
+
+// DeleteDaemonset makes a best effort at deleting a daemonset and its pods, then waits for them to be deleted
+func DeleteDaemonset(context *clusterd.Context, namespace, name string) {
+	logger.Infof("removing %s daemonset if it exists", name)
+	deleteAction := func(name string, options *metav1.DeleteOptions) error {
+		return context.Clientset.ExtensionsV1beta1().DaemonSets(namespace).Delete(name, options)
+	}
+	getAction := func(name string) error {
+		_, err := context.Clientset.ExtensionsV1beta1().DaemonSets(namespace).Get(name, metav1.GetOptions{})
+		return err
+	}
+	deletePodsAndWait(context, namespace, name, deleteAction, getAction)
+}
+
+// deletePodsAndWait will delete a resource, then wait for it to be purged from the system
+func deletePodsAndWait(context *clusterd.Context, namespace, name string,
+	deleteAction func(string, *metav1.DeleteOptions) error,
+	getAction func(string) error) {
 
 	var gracePeriod int64
 	propagation := metav1.DeletePropagationForeground
 	options := &metav1.DeleteOptions{GracePeriodSeconds: &gracePeriod, PropagationPolicy: &propagation}
 
 	// Delete the deployment if it exists
-	err := context.Clientset.ExtensionsV1beta1().Deployments(namespace).Delete(name, options)
+	err := deleteAction(name, options)
```

Comment:
You can just called deleteAction() without passing any args since you already did that in the body of the annonymous function.